### DISCUSSION TITLE
⏳ 🔗 Add Cloudflare Turnstile integration

### DISCUSSION THREAD

**Discussion by reedock** [elementor/elementor#25671] [Feature Request]
### Prerequisites

- [X] I have searched for similar feature requests in both open and closed discussions and cannot find an existing request.
- [X] I have verified that the feature is still missing from the latest stable versions of Elementor or Elementor Pro.

### What problem is your feature request going to solve? Please describe..

There is currently no way to add Cloudflare Turnstile to Elementor forms without the need of an extra plugin. Only reCAPTCHA is being offered.

### Describe the solution you'd like.

Add Cloudflare Turnstile in the Settings, Integration tab.

### Describe alternatives you've considered.

reCAPTCHA and reCAPTCHA V3 aren't as good as Cloudflare Turnstile in my humble opinion.

### Additional context

Other plugins specialized in "forms" like WPforms etc. integrated Cloudflare turnstile already. 

### Agreement

- [X] I agree that my request may be closed without action if it doesn't meet all the requirements.

**Comment by reedock** [comment]
Hope this can be added like WPForms did so it can be integrated easily to Elementor forms.

**Comment by liv2day** [comment]
This already exists in https://wordpress.org/plugins/simple-cloudflare-turnstile/ right..? and duplicate of https://github.com/orgs/elementor/discussions/27293

**Reply by reedock** [reply]
"There is currently no way to add Cloudflare Turnstile to Elementor forms **without the need of an extra plugin**. Only reCAPTCHA is being offered."

And for the duplicate, my post was made on April 9 and https://github.com/orgs/elementor/discussions/27293 was created May 10.

**Reply by nicocart** [reply]
integration but extra plugin

**Reply by reedock** [reply]
This feature request is clear. Integrate Cloudflare Turnstile directly to Elementor settings so there is no need for a plugin.

**Comment by moonbeetle** [comment]
At this moment (Dec 4th 2024), the Simple Cloudflare Turnstile throws a JS error Can't find variable: turnstile on websites with Elementor.  So an integration with Cloudflare Turnstile without the need for a plugin would make it a more reliable solution. Also beneficial for Elementor.

**Comment by jiimmyy1** [comment]
Very necessary to have this in the settings of Elementor without installing a plugin. Turnstile is 100000 times better. Please, is very important!
Thanks in advance

**Comment by DavePodosyan** [comment]
I built a custom integration to add Cloudflare Turnstile support to Elementor Forms, following the same pattern Elementor uses for reCAPTCHA. To use it, you just need to include it in your theme's function.php file. 

https://gist.github.com/DavePodosyan/b4e6f0a261ce5c7ed3b30b0734d56291

**Comment by robwent** [comment]
This will become more important this year as Google are moving all keys to cloud projects (Billing) by the end of the year and lots of users are going to want an alternative.

hcaptcha is another option that could be added.

**Comment by ghost** [comment]
I need this on over 50 websites!

**Comment by freshinkcooperative** [comment]
Google will start charging for ReCaptcha at the end of 2025, it would be very important for a lot of people to get this before then.

**Comment by reedock** [comment]
reCAPTCHA Essentials is only free up to 10,000 assessments per month. 

That might sound like a lot but good luck if you have many visitors and bots monthly...

Source:
https://cloud.google.com/recaptcha/docs/billing-information

**Comment by rosskevin** [comment]
Thanks @DavePodosyan for the starter implementation.  Note that your implementation uses `implicit` mode and does not work for forms that are contained in popups.  The `implicit` initialization will execute, the `cf-turnstile` is not found, which leads to `[Cloudflare Turnstile] Error: 300030` and `Turnstile Widget seem to have hung`.  I'm not complaining @DavePodosyan! I just wanted to share more implementation information that can solve this.

The solution means switching the cloudflare turnstile to `explicit`, and using a js `MutationObserver` to watch for forms that pop in.  But, I just discovered that while I got this to work, opening then closing, then opening again failed to re-init the turnstile.

Then I found a related issue for a different modal/popup:
https://community.cloudflare.com/t/turnstile-not-working-inside-fancybox-modal/562481

Essentially, track state of initilization on the turnstile field element itself as an attribute, and leave `MutationObserver`s connected for the pop/close/pop scenario to re-init.

We have our own custom plugin (not commercial), so my implementation has been very much customized, but I'd be happy to share code if that's of interest...it could be cleaned up.

So, a universal implementation:
- needs to use turnstile `explicit` mode
- needs to reconnect `MutationObserver`s to track multiple potential popup forms (as well as pop/close/pop scenario).  This should reconnect via the `render(callback)` in case user interaction occurs
- needs to track init/rendered state on the html element (pop/close/pop scenario)
- needs to track init/rendered id's on page state (memory) so that before re-render, a `turnstile.remove()` is executed

I hope this info is useful to others that might be trying and encountering `[Cloudflare Turnstile] Error: 300030` and `Turnstile Widget seem to have hung`

**Reply by DavePodosyan** [reply]
Thank you so much @rosskevin for your detailed explanation of the issue my solution was facing. I hadn’t tested it with popups, so I wasn’t even aware this could break things like that.

I’ve now updated the gist to use explicit Turnstile rendering, and it’s still a single-file solution. Please have a look, would love to hear if it works as expected now. 

https://gist.github.com/DavePodosyan/b4e6f0a261ce5c7ed3b30b0734d56291

**Comment by deepwaterpixel** [comment]
The lack of Turnstile support has caused me to stop using Elementor forms altogether.  There are too many other options that support Turnstile and hCaptcha.

**Comment by spasauna** [comment]
+1

**Comment by TomerBu** [comment]
+1

**Comment by DavePodosyan** [comment]
A lightweight plugin that adds native support for hCaptcha and Cloudflare Turnstile

https://github.com/DavePodosyan/captcha-for-elementor-pro-forms

**Reply by reedock** [reply]
Thank you for this but I hope you don't continue promoting your plugin here as my request was to avoid using **AN EXTRA** plugin.

**What problem is your feature request going to solve? Please describe..**
There is currently no way to add Cloudflare Turnstile to Elementor forms without the need of an extra plugin. Only reCAPTCHA is being offered.

**Comment by DavePodosyan** [comment]
@reedock, Totally get it, native support in Elementor would be ideal. 

I’m not trying to promote anything here, just leaving a stopgap for folks blocked today. The add-on is minimalist: it registers the field types, enqueues the provider scripts, and validates tokens on submit. No settings beyond site/secret keys, no extra features, no telemetry.

I would have submitted this as a PR directly, but unfortunately Elementor Pro is not a public repo.

**Reply by reedock** [reply]
Yet instead of just replying to the comment, you posted another fresh one and listed all the good things your plugin does.

It gives off an “I’m not trying to get you to eat this cake since I know you don’t like them, but it’s fresh out of the oven, packed with chocolate, and you can have it for free” kind of vibe...

Thanks for not hijacking this feature request to promote your plugin even if he does wonder.

**Reply by DavePodosyan** [reply]
The details were to clarify that it’s a minimal workaround, not a bloated add-on, and to explain why it can’t be submitted as a PR.
If you or others don’t want a stopgap, that’s perfectly fine, but implying I’m trying to sneak in promotion here is a misread. I’m here contributing to the discussion and offering an immediate path for those who need it now. Thanks

**Reply by reedock** [reply]
Thank you for your contribution 👍

**Reply by DavePodosyan** [reply]
Appreciate it 👍  Hopefully Elementor makes it native soon so we don't have to eat these fresh-out-of-the-oven cakes anymore.

**Comment by caleb-bright** [comment]
I would love to see this in v4.

**Comment by MorrowShore** [comment]
any news on this?

**Comment by marowi85** [comment]
+1 would be a nice addition

**Comment by gallnomad** [comment]
I would also like to see this in Elementor without having to use a plugin.

**Comment by SteinBC** [comment]
+1

**Comment by benrk** [comment]
This is an important intergration. +1

**Comment by R-Adrian** [comment]
+1, Cloudflare Turnstile (free) is needed as a replacement for reCAPTCHA (free) because Google is forcing all current users of reCaptcha classic to migrate to the new reCaptcha on Google Cloud by the end of 2025, and that is only free for only 10.000 visits - good luck with that when you have a ton of AI bots crawling your site.

if you have seen any reCaptcha rendered in the past week, you probably noticed that the reCaptcha button has a new text overlay that announces the change of terms of service and this link for "Take action":  https://google.com/recaptcha/admin/migrate

<img width="326" height="104" alt="image" src="https://github.com/user-attachments/assets/4906bdc4-560f-4703-bd4a-e51992148943" />


https://cloud.google.com/recaptcha/docs/migration-overview

quote from that Google overview reference:
Migration timeline

All reCAPTCHA Classic customers must migrate or will be migrated to Google Cloud by the end of 2025. The process involves the following phases.

Q3 2024: New Classic keys are no longer allowed.
Q1 2025: The first notification that informs customers to migrate using the self-migration tool is sent.
Q3 2025: A follow-up notification is sent to customers that haven't yet migrated from Classic, which links directly to the [migration tool](https://www.google.com/recaptcha/admin/migrate) in the [Google Cloud console](https://cloud.google.com/recaptcha/docs/migrate-recaptcha#migrate-through-cloud-console).

**Q4 2025: Customers are automatically migrated from Classic to the Enterprise version.
Q4 2025: All Classic keys are rejected, with API access locked for keys without a Google Cloud project.**

**Reply by MorrowShore** [reply]
Very good point.

**Comment by joaopedrofrech** [comment]
+1 for this request. Cloudflare Turnstile is necessary.

**Comment by deepwaterpixel** [comment]


**Reply by deepwaterpixel** [reply]
I have removed all of my comments.

**Comment by MorrowShore** [comment]
What's the usual turn around of popular feature requests?

**Reply by SteinBC** [reply]
Wondering this too

**Reply by reedock** [reply]
Would be great to have someone from Elementor team to show up and officially say if they will implement it or no at this point. I rather have someone to lock this topic and say no than hoping for more years like Font Awesome...

**Reply by jiimmyy1** [reply]
Hope the staff answers

**Reply by gallnomad** [reply]
I wrote to them on 16th October 2025 and this was the reply

Thank you for contacting Elementor Support. I'm Trina, a member of the Elementor Technical Support team. I will be helping you with your issue today.

After consulting on this, it appears that this feature request has already been documented for our developers. While we don’t have an ETA to share at the moment, you can keep an eye on our changelogs for any related updates from our development team:
>>>> [Elementor Changelogs](https://elementor.com/help/elementor-products-changelogs/)

You can also add your vote and share feedback in this discussion thread to help prioritize the request:
>>> [GitHub Discussion](https://github.com/orgs/elementor/discussions/25671)

Apologies for not being able to provide direct assistance, and thank you for your patience and understanding!

If you have any further questions about Elementor and its current features, please do not hesitate to contact us. We will be more than happy to assist.

Regards,

**Reply by MorrowShore** [reply]
> I wrote to them on 16th October 2025 and this was the reply

Thank you for looking into it.

**Comment by lovepixel-git** [comment]
+1

**Comment by codeagencybe** [comment]
Wow, 1.5 years later still no turnstile while the entire industry is using this as a standard. 

Shame on you Elementor

**Reply by teoeugene** [reply]
It's typical of Elementor. Highly requested features in the past were opened for 4-6 years. And with V4 high in their priority, this will probably be in the backlog for a long time.

**Comment by gallnomad** [comment]
This is how pathetic this is on Elementor's part. Two months ago, I asked a small plugin developer of WP24 Domain Check if they had any plans of putting Turnstile in, as I no longer use Google's Captcha. I didn't hold high hopes as it is a free plugin from a small developer. Today, it was rolled out in their latest version, **TWO MONTHS** after asking. If a small developer of a free plugin can manage to make it a priority, not sure as a paying customer of Elementor they can't take this more seriously. I don't think it is even on the development list!

**Reply by reedock** [reply]
What I cannot understand is the lack of official intervention from Elementor to clearly state whether this can or cannot be implemented, or to explain why it will not be implemented and simply close the request instead of giving false hope. It is pretty useless to make new requests if even the top ones are not being looked at...

**Reply by bclements-northpinecreative** [reply]
I just tried two of the free solutions that others have developed, and neither worked properly. Either stopped the form from submitting, or would not work if you reloaded the page. I tried fixing the code, but without any luck.

To your point, it would be nice if they just addressed it and said they are working with Cloudflare on a solution, but maybe they are running into the same issues I am.

**Reply by reedock** [reply]
I hope I am wrong and that you are right, but I really doubt they have been working with Cloudflare for almost two years on a solution. If they were, it would be smart to take 30 seconds to say here that it is being worked on. That would avoid users getting frustrated. Turnstile has been implemented by almost everyone except Elementor.

**Reply by endymion0000** [reply]
> I just tried two of the free solutions that others have developed, and neither worked properly. Either stopped the form from submitting, or would not work if you reloaded the page. 

I've used Simple CAPTCHA Alternative with Cloudflare Turnstile and maybe this will work if you have another plugin, but select the form in the editor, go to Advanced and in the Layout section, set Cache Settings to Inactive.

This fixed things for me when I had issues.

**Reply by danielgutierrezcolorado** [reply]
thx for the tip of disabling cache on the form!

**Comment by caleb-bright** [comment]
Really hoping this comes in 2026.

**Comment by bclements-northpinecreative** [comment]
It is really odd to me that they mention Cloudflare so many times on their WordPress hosting page on their website. Cloudflare CDN is mentioned 8 times, if I counted correctly. So, if they are so proud to partner with them, why is this integration taking so long? Now is the time to launch this feature; everyone is moving away from reCAPTCHA!

Please at least tell us it is in the product roadmap. Paying elementor customer here!

**Comment by MorrowShore** [comment]
Get this thing added already. It's way more important than whatever AI slop is being worked on right now. It's an actual feature that actual site owners need.

**Comment by onan** [comment]
Huge value for Elementor if this feature gets added. Current free implementations on Wordpress plugins don't work well with Elementor Forms. Please consider.

**Comment by tomm1079** [comment]
agree would love for this to be added

**Comment by gjportegies** [comment]
Would be very nice to have this integrated. Other themes like Avada already have this integrated and Google is forcing their terrible Google Cloud Console platform.

**Comment by TOTLD** [comment]
Is this even in the roadmap? Any update at all?

CF Turnstile has been a free public product since 2022...

As a paying customer I am increasingly worried that Elementor overall is heading in the totally wrong direction. Even the paid 'support' is woeful now unless you upgrade to a specific tier.

FYI anyone here thinking of alternatives, Bricks has supported Turnstile since v1.9 and continues to do so, because it is clearly a feature that any form-builder should include.

**Reply by gallnomad** [reply]
No it isn't. According to the email I got from them this morning.

**Comment by reedock** [comment]
This request is nearing the 2-year mark… Not sure what’s the point of feature requests if they aren’t addressed by Elementor staff. It’s in the top 10 all time requested features for a while now. So frustrating to see requests sit for years while they keeps working on promoting new paid products instead of addressing top features and bugs.

**Reply by jiimmyy1** [reply]
totally agree. wtf is going on? @KingYes with all respect, please take care of your customers and release important stuff that is asked not stupid shit that nobody will use

**Reply by gjportegies** [reply]
In addition to this; it's also a pretty easy integration.

**Reply by gallnomad** [reply]
I emailed them last year and again this morning. This is the reply this morning. What a joke!

Thank you for contacting Elementor Support.
I'm Kevin, and I will assist you today.

Thank you for following up on this, and I sincerely apologize for the lack of updates and the frustration this has caused. We completely understand how important Cloudflare Turnstile support is, especially with the recent changes to CAPTCHA services.

At this time, there are no confirmed updates we can share regarding its implementation. However, please be assured that your request will be formally logged as a feature request and shared with our product team for consideration.

We truly value your long-time support and feedback, and we understand your concerns about this functionality. Thank you for bringing this to our attention, and we appreciate your patience.

**Comment by gallnomad** [comment]
I suggest if you want this done you bombard their support with emails. I emailed them last year and again this morning. This mornings was essentially the same as the one I got last year (see below). It is a complete joke. I emailed a really small plugin developer and asked if they were going to implement Turnstile. 2 months later it was implemented and this was on a free plugin. I asked Elementor why I'm paying them when I small free plugin can manage to do it.



**Thank you for contacting Elementor Support.
I'm Kevin, and I will assist you today.

Thank you for following up on this, and I sincerely apologize for the lack of updates and the frustration this has caused. We completely understand how important Cloudflare Turnstile support is, especially with the recent changes to CAPTCHA services.

At this time, there are no confirmed updates we can share regarding its implementation. However, please be assured that your request will be formally logged as a feature request and shared with our product team for consideration.

We truly value your long-time support and feedback, and we understand your concerns about this functionality. Thank you for bringing this to our attention, and we appreciate your patience.

Best Regards,**

**Reply by gallnomad** [reply]
Follow up from Elementor - Thank you for sharing your thoughts, and I completely understand your frustration. I sincerely apologize for the delay in implementing Cloudflare Turnstile and for the disappointment this has caused, especially given your long-time support.

While the feature may seem straightforward, its implementation requires careful integration across multiple Elementor products and compatibility with various site setups, which can take longer than anticipated. Please rest assured that your request has been noted, and we continue to monitor and prioritize feature requests like this for future updates.

We truly appreciate your patience and understanding.

**Reply by deepwaterpixel** [reply]
This response from them is crazy simply because this thread exists.  Literally people posting code freely to enable it in Elementor and they're just like "oh no it's so hard it takes long time boohoo".

Pathetic response from support.  I wouldn't have sent this response to a customer if you had me at gunpoint, you'd have to get someone else to send it for me because WUT

Why can't they just be honest: "We don't care about improving existing features for already paying customers because we have all of our teams working solely on revenue generation and not on quality control."

**Reply by joaopedrofrech** [reply]
Elementor is a lost cause

**Comment by salmanjaveed** [comment]
+1

**Comment by joaopedrofrech** [comment]
They're probably too busy worrying about revenue with their new editor and useless stuff. Why bother working on improvements and features behind the scenes? That's why I only use it for legacy projects. With AI tools on the field, no-code tools and way less necessary every day.

**Reply by DavePodosyan** [reply]
What do you use for new projects?

**Reply by joaopedrofrech** [reply]
> What do you use for new projects?

I use Nextjs as framework and Payload as CMS. Claude Code with Antigravity connected with Stitch and Penpot/Figma MCP servers. 

Stitch helps to generate UI ideas and elements, the ai can send it to Figma or Penpot via mcp, then when it's ready you send it back to your project as react components, after frontend ready, claude can help with the backend with next and payload.

**Reply by DavePodosyan** [reply]
Thank you for your detailed reply. Much appreciated ✨

**Comment by narinder9** [comment]
Hi everyone,

I have been following this discussion and noticed many of you are looking for a reliable way to use **Cloudflare Turnstile with Elementor Forms** without manual coding.

To help solve this, we have integrated native Cloudflare Turnstile support into our plugin, **CoolFormKit**. It allows you to replace reCAPTCHA with **Turnstile** directly within the Elementor **Form widget** settings.

You can check out the integration details here:

* **Setup Guide:** How to add Cloudflare Turnstile to Elementor :- https://coolformkit.com/features/cloudflare-turnstile-for-elementor-form/?utm_source=github_issue

We're constantly improving our integrations, so if you try it out and have any feedback,please let me know!

**Reply by reedock** [reply]
This feature request is clear. Integrate Cloudflare Turnstile directly to Elementor settings so there is no need for a plugin. Thank you for not using this place for promoting another paid product or a plugin...

**Comment by jiimmyy1** [comment]
Doesn’t work well with Elementor Forms

**Comment by jiimmyy1** [comment]
I tried it multiple times, but console gives error.
Also in popups doesnt load well…

**Comment by reedock** [comment]
This feature request is clear. Integrate Cloudflare Turnstile directly to Elementor settings so there is no need for a plugin. Thank you for not using this place for promoting another paid product or a plugin...

**Comment by reedock** [comment]
Today marks 2 years since this feature request was created. It’s one of the most requested features and there’s still no sign from Elementor’s team...

**Reply by caleb-bright** [reply]
I've talked to people from Elementor and mentioned this feature request. I was told they are looking at different spam protection options and they will consider adding a Turnstile integration in a future release. I'm hoping it's added by the end of this year.

**Reply by joaopedrofrech** [reply]
Well, it's been already two years, what is one more, right? And I don't doubt this new spam protection system they're building is some sort of subscription system, as everything they're building recently.

**Reply by caleb-bright** [reply]
Ultimately, time will tell.

**Reply by reedock** [reply]
Elementor form is still not compatible with [reCAPTCHA Enterprise](https://github.com/orgs/elementor/discussions/33466) so what “spam protection” are left to consider at this point? I hope they are not considering **AGAIN** adding another product behind credits...

**Comment by reedock** [comment]
<img width="545" height="580" alt="facebook-post" src="https://github.com/user-attachments/assets/ef87d131-3d6f-460e-b167-ee97ff524f6e" />

Simply trying to ask on the Facebook group if it’s still worth making new feature requests knowing they are left unattended for years will trigger the moderator to give you a warning and remove your post… lol!

**Comment by spasauna** [comment]
Generally speaking and not only related to this feature request, but why is there such a bad communication on Github regarding feature requests and bug reports? Most people here even got Elementor Pro and are willing to sacrifice time and put in great effort to sometimes write detailed feature requests and bug reports just to get ignored.

That should NOT be the case. I even got a bug report back from 2024 that seems to be completely unbothered.

**Reply by gallnomad** [reply]
They don't seem particularly bothered when you email them and call them out on this.

**Reply by gursuj** [reply]
Might be due to the large volume of opened issues here. I assume they just prioritize support channels for pro users or their own internal roadmaps (à la the push for Atomic Elements). Will say, github would be a way better platform to communicate openly if they  chose to do so.

**Reply by gursuj** [reply]
This just in: [Elementor cuts 100 jobs, 30% of workforce](https://www.calcalistech.com/ctechnews/article/sycgn6yxze)

**Comment by Jarpiino** [comment]
up.

**Comment by gjportegies** [comment]
Up

**Comment by Yorkki** [comment]
+1

Already installed few Avada's, and as long as Elementor is forcing me to do that, that's the way it's gonna go. If I get too comfy with it, I might even move to it, permanently. I'm not really getting it why they won't do this, IT IS NEEDED.

**Comment by tonydjukic** [comment]
So this is frustrating.  Came here to see if this was a feature request or not and if not, add it myself... ...only to discover that it's existed for 2+ years already and is completely ignored.  The worst part is that Google is starting to push people away from v3 in favor of their enterprise version so even what already exists in Elementor as an integration is headed towards obsolescence.

**Comment by Jarpiino** [comment]
Money money money

**Comment by reedock** [comment]
Cloudflare Turnstile has been in the Top 5 most requested features of all time for a while since the feature request below has been "locked" 😉..
[Ability to disable features (Announcement Popups, Elementor AI)](https://github.com/orgs/elementor/discussions/22796)
